frame/core-fellowship/src/lib.rs

```diff
@@ -0,0 +1,597 @@
+					Event::<T, I>::Demoted { who, to_rank }
+				} else {
+					Member::<T, I>::remove(&who);
+					Event::<T, I>::Offboarded { who }
+				};
+				Self::deposit_event(event);
+				return Ok(Pays::No.into())
+			}
+
+			Err(Error::<T, I>::NothingDoing.into())
+		}
+
+		/// Set the parameters.
+		///
+		/// - `origin`: A origin complying with `ParamsOrigin` or root.
+		/// - `params`: The new parameters for the pallet.
+		#[pallet::weight(T::WeightInfo::set_params())]
+		#[pallet::call_index(1)]
+		pub fn set_params(
+			origin: OriginFor<T>,
+			params: Box<ParamsOf<T, I>>,
+		) -> DispatchResultWithPostInfo {
+			T::ParamsOrigin::try_origin(origin).map(|_| ()).or_else(|o| ensure_root(o))?;
+			Params::<T, I>::put(params.as_ref());
+			Self::deposit_event(Event::<T, I>::ParamsChanged { params: *params });
+			Ok(Pays::No.into())
+		}
+
+		/// Set whether a member is active or not.
+		///
+		/// - `origin`: A `Signed` origin of a member's account.
+		/// - `is_active`: `true` iff the member is active.
+		#[pallet::weight(T::WeightInfo::set_active())]
+		#[pallet::call_index(2)]
+		pub fn set_active(origin: OriginFor<T>, is_active: bool) -> DispatchResult {
+			let who = ensure_signed(origin)?;
+			ensure!(
+				T::Members::rank_of(&who).map_or(false, |r| !r.is_zero()),
+				Error::<T, I>::Unranked
+			);
+			let mut member = Member::<T, I>::get(&who).ok_or(Error::<T, I>::NotTracked)?;
+			member.is_active = is_active;
+			Member::<T, I>::insert(&who, &member);
+			Self::deposit_event(Event::<T, I>::ActiveChanged { who, is_active });
+			Ok(())
+		}
+
+		/// Approve a member to continue at their rank.
+		///
+		/// This resets `last_proof` to the current block, thereby delaying any automatic demotion.
+		///
+		/// If `who` is not already tracked by this pallet, then it will become tracked.
+		/// `last_promotion` will be set to zero.
+		///
+		/// - `origin`: An origin which satisfies `ApproveOrigin` or root.
+		/// - `who`: A member (i.e. of non-zero rank).
+		/// - `at_rank`: The rank of member.
+		#[pallet::weight(T::WeightInfo::approve())]
+		#[pallet::call_index(3)]
+		pub fn approve(
+			origin: OriginFor<T>,
+			who: T::AccountId,
+			at_rank: RankOf<T, I>,
+		) -> DispatchResultWithPostInfo {
+			match T::PromoteOrigin::try_origin(origin) {
+				Ok(allow_rank) => ensure!(allow_rank >= at_rank, Error::<T, I>::NoPermission),
+				Err(origin) => ensure_root(origin)?,
+			}
+			ensure!(at_rank > 0, Error::<T, I>::InvalidRank);
+			let rank = T::Members::rank_of(&who).ok_or(Error::<T, I>::Unranked)?;
+			ensure!(rank == at_rank, Error::<T, I>::UnexpectedRank);
+			let mut member = Member::<T, I>::get(&who).ok_or(Error::<T, I>::NotTracked)?;
+
+			member.last_proof = frame_system::Pallet::<T>::block_number();
+			Member::<T, I>::insert(&who, &member);
+
+			Self::dispose_evidence(who.clone(), at_rank, Some(at_rank));
+			Self::deposit_event(Event::<T, I>::Proven { who, at_rank });
+
+			Ok(Pays::No.into())
```

Comment:
Same comment about aggressive `Pays::No` here. It is not clear to me that there are any limits to how many times a valid origin can make this call, and thus it is vulnerable to attack.

Reply:
and below too, won't repeat myself.